libnmstate/state.py

```diff
@@ -276,64 +276,41 @@ def merge_interfaces(self, other_state):
             dict_update(other_state.interfaces[name], self.interfaces[name])
             self._ifaces_state[name] = other_state.interfaces[name]
 
-    def merge_route_config(self, other_state):
+    def merge_routes(self, other_state):
         """
-        Converting route partial editing to route full editing by
-        merging config routes from other state to self.
-        Assuming the other state is from `netinfo.show()` which don't have
-        absent routes in it.
+        Given the self and other states, complete the self state by merging
+        the routes form the current state.
+        The resulting routes are a combination of the current routes and the
+        desired routes:
+        - Self routes are kept.
+        - Other routes are kept (if any routes exist in iface self).
+        - Self absent routes overwrite other routes state.
+          - Support wildcard for matching the absent routes.
         """
-        iface_enable_states = _get_iface_enable_states(self, other_state)
-        ipv4_enable_states = _get_ip_enable_states(
-            Interface.IPV4, self, other_state)
-        ipv6_enable_states = _get_ip_enable_states(
-            Interface.IPV6, self, other_state)
-        self_iface_route_sets = defaultdict(set)
-        absent_route_sets = set()
-        other_iface_route_sets = defaultdict(set)
-
-        for route in self._config_routes:
-            if route.get(Route.STATE) == Route.STATE_ABSENT:
-                absent_route_sets.add(RouteEntry(route))
-
-        for route in other_state._config_routes:
-            # The other_state(running state) does not have any absent
-            # route entry and all have NEXT_HOP_INTERFACE
-            iface_name = route[Route.NEXT_HOP_INTERFACE]
-            other_iface_route_sets[iface_name].add(RouteEntry(route))
+        other_routes = set()
+        for ifname, routes in six.viewitems(other_state.config_iface_routes):
+            if self.config_iface_routes.get(ifname):
+                other_routes |= set(routes)
```

Comment:
In https://github.com/nmstate/nmstate/pull/384/commits/c26b0ecd457a5431d93a8b16dee480198b35fbc1#diff-6f12cd36ee493d92b4777db9ba2780feR184 you support routes referring to non-defined interface.
But here you discard any current route to interface not found in desire state.